src/cache.c

```diff
@@ -0,0 +1,388 @@
+    if (!cache)
+        return NULL;
+    cache_size_bits = size_bits;
+    cache_size = 1 << size_bits;
+
+    for (int i = 0; i < N_CACHE_LIST_TYPES; i++) {
+        cache->list_table[i] = malloc(sizeof(struct list_head));
+        INIT_LIST_HEAD(cache->list_table[i]);
+        cache->list_size[i] = 0;
+    }
+
+    cache->map = malloc(sizeof(hashtable_t));
+    if (!cache->map) {
+        free(cache->list_table);
+        free(cache);
+        return NULL;
+    }
+    cache->map->ht_list_head = malloc(cache_size * sizeof(struct hlist_head));
+    if (!cache->map->ht_list_head) {
+        free(cache->map);
+        free(cache->list_table);
+        free(cache);
+        return NULL;
+    }
+    for (uint32_t i = 0; i < cache_size; i++) {
+        INIT_HLIST_HEAD(&cache->map->ht_list_head[i]);
+    }
+
+    cache->capacity = cache_size;
+    cache->lru_capacity = cache_size / 2;
+    return cache;
+}
+
+/* Rule of ARC
+ * 1. size of LRU_list + size of LFU_list <= c
+ * 2. size of LRU_list + size of LRU_ghost_list <= c
+ * 3. size of LFU_list + size of LFU_ghost_list <= 2c
+ * 4. size of LRU_list + size of LFU_list + size of LRU_ghost_list + size of
+ * LFU_ghost_list <= 2c
+ */
+#define CACHE_ASSERT(cache)                                                 \
+    assert(cache->list_size[LRU_list] + cache->list_size[LFU_list] <=       \
+           cache->capacity);                                                \
+    assert(cache->list_size[LRU_list] + cache->list_size[LRU_ghost_list] <= \
+           cache->capacity);                                                \
+    assert(cache->list_size[LFU_list] + cache->list_size[LFU_ghost_list] <= \
+           2 * cache->capacity);                                            \
+    assert(cache->list_size[LRU_list] + cache->list_size[LRU_ghost_list] +  \
+               cache->list_size[LFU_list] +                                 \
+               cache->list_size[LFU_ghost_list] <=                          \
+           2 * cache->capacity);
+
+static inline void move_to_mru(cache_t *cache,
+                               arc_entry_t *entry,
+                               const cache_list_t arc_type)
+{
+    cache->list_size[entry->arc_type]--;
+    cache->list_size[arc_type]++;
+    entry->arc_type = arc_type;
+    list_del_init(&entry->list);
+    list_add(&entry->list, cache->list_table[arc_type]);
+}
+
+static inline void replace_list(cache_t *cache)
+{
+    if (cache->list_size[LRU_list] >= cache->lru_capacity)
+        move_to_mru(
+            cache,
+            list_last_entry(cache->list_table[LRU_list], arc_entry_t, list),
+            LRU_ghost_list);
+    else if (cache->list_size[LFU_list] >=
+             (cache->capacity - cache->lru_capacity))
+        move_to_mru(
+            cache,
+            list_last_entry(cache->list_table[LFU_list], arc_entry_t, list),
+            LFU_ghost_list);
+}
+
+void *cache_get(cache_t *cache, uint32_t key)
+{
+    if (cache->capacity <= 0 ||
```

Comment:
The type of `capacity` is `uint32_t`, and it is always >= 0. Therefore, you shall never validate it with `<= 0`.

Reply:
I modify it to 
``` c
if (!cache->capacity)
```